File: ObscuritasMediaManager.Client/ObscuritasMediaManager.Client/Extensions/MediaFilterService.cs
using ObscuritasMediaManager.Client.BusinessComponents.MediaFilter;

public class MediaFilterService
{
    public List<MediaModel> filter(List<MediaModel> result, MediaFilter filter)
    {
        /* TODO ObjectFilterExtensions.applyArrayFilter(result, filter.genres, x => x.Genres); */

        result = result.applyPropertyFilter(filter.Ratings, x => x.Rating)
            .applyMultiPropertySearch(filter.Search ?? string.Empty, x => x.Name, x => x.Description)
            .applyPropertyFilter(filter.Status, x => x.Status)
            .applyRangeFilter(filter.Release, x => x.Release)
            .applyPropertyFilter(filter.Languages, x => x.Language)
            .applyPropertyFilter(filter.Category, x => x.Type)
            .applyArrayFilter(filter.ContentWarnings, x => x.ContentWarnings)
            .applyPropertyFilter(filter.TargetGroups, x => x.TargetGroup, CheckboxState.Ignore)
            .ToList();

        if (filter.SortingPropertyExpression is null) return result;
        if ((filter.SortingDirection is null) || (filter.SortingDirection == SortDirection.Ascending))
            return result.OrderBy(x => filter.SortingPropertyExpression.Compile()(x)).ToList();
        else
            return result.OrderByDescending(x => filter.SortingPropertyExpression.Compile()(x)).ToList();
    }
}

File: Frontend/services/media-filter.service.js
import { MediaFilter } from '../advanced-components/media-filter-sidebar/media-filter.js';
import { CheckboxState } from '../data/enumerations/checkbox-state.js';
import { MediaModel } from '../obscuritas-media-manager-backend-client.js';
import { sortBy } from './extensions/array.extensions.js';
import { ObjectFilterService } from './object-filter.service.js';

export class MediaFilterService {
    /**
     *
     * @param {MediaModel[]} result
     * @param {MediaFilter} filter
     */
    static filter(result, filter) {
        ObjectFilterService.applyPropertyFilter(result, filter.ratings, 'rating');
        ObjectFilterService.applyArrayFilter(result, filter.genres, 'genres', (x) => x.id);
        ObjectFilterService.applyMultiPropertySearch(
            result,
            filter.search ?? '',
            'name',
            'description',
            'kanjiName',
            'romajiName',
            'germanName',
            'englishName'
        );
        ObjectFilterService.applyPropertyFilter(result, filter.status, 'status');
        ObjectFilterService.applyRangeFilter(result, filter.release, 'release');
        ObjectFilterService.applyPropertyFilter(result, filter.languages, 'language');
        ObjectFilterService.applyPropertyFilter(result, filter.category, 'type');
        ObjectFilterService.applyArrayFilter(result, filter.contentWarnings, 'contentWarnings');
        ObjectFilterService.applyPropertyFilter(result, filter.targetGroups, 'targetGroup', CheckboxState.Ignore);
        ObjectFilterService.applyValueFilter(result, filter.deleted, 'deleted');
        ObjectFilterService.applyValueFilter(result, filter.complete, 'complete');

        if (!filter.sortingProperty) return result;
        var sorted = sortBy(result, (x) => x[filter.sortingProperty]);
        if (!filter.sortingDirection || filter.sortingDirection == 'ascending') return sorted;
        else return sorted.reverse();
    }
}
